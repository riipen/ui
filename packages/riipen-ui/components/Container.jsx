import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";

class Container extends React.Component {
  static propTypes = {
    /**
     * The content inside the container.
     */
    children: PropTypes.any,

    /**
     * List of additional classes to apply to this component.
     */
    classes: PropTypes.array,

    /**
     * Determine the max-width of the container.
     * The container width grows with the size of the screen.
     */
    maxWidth: PropTypes.oneOf(["sm", "md", "lg", "xl", false])
  };

  static defaultProps = {
    classes: [],
    maxWidth: "lg"
  };

  static contextType = ThemeContext;

  render() {
    const { classes, children, maxWidth } = this.props;

    const theme = this.context;

    const className = clsx(classes);

    return (
      <React.Fragment>
        <div className={className}>{children}</div>
        <style jsx>{`
          div {
            margin: 0 auto;
            max-width: ${maxWidth
              ? `${theme.breakpoints[maxWidth]}px`
              : "100%"};
            position: relative;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default Container;
