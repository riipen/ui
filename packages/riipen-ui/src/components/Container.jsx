import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import ContainerHeader from "./ContainerHeader";

class Container extends React.Component {
  static displayName = "Container";

  static propTypes = {
    /**
     * Whether to have a border around the container.
     */
    border: PropTypes.bool,

    /**
     * The content inside the container.
     */
    children: PropTypes.any,

    /**
     * List of additional classes to apply to this component.
     */
    classes: PropTypes.array,

    /**
     * The color of the container.
     */
    color: PropTypes.oneOf("default", "white"),

    /**
     * The props to pass to the ContainerHeader.
     */
    headerProps: PropTypes.object,

    /**
     * Determine the max-width of the container.
     * The container width grows with the size of the screen.
     */
    maxWidth: PropTypes.oneOf(["sm", "md", "lg", "xl", false])
  };

  static defaultProps = {
    border: false,
    classes: [],
    color: "default",
    maxWidth: "lg"
  };

  static contextType = ThemeContext;

  render() {
    const {
      border,
      classes,
      children,
      color,
      headerProps,
      maxWidth
    } = this.props;

    const theme = this.context;

    const className = clsx("root", classes);

    return (
      <React.Fragment>
        <div className={className}>
          {headerProps && <ContainerHeader {...headerProps} />}
          <div className={clsx(border && "border", color)}>{children}</div>
        </div>
        <style jsx>{`
          .root {
            box-sizing: border-box;
            margin: 0 auto;
            max-width: ${maxWidth
              ? `${theme.breakpoints[maxWidth]}px`
              : "100%"};
            padding: 0 ${theme.spacing(6)}px;
            position: relative;
          }

          .border {
            box-shadow: ${theme.shadows[1]};
          }

          .white {
            background-color: ${theme.palette.common.white};
          }

          @media (max-width: ${theme.breakpoints.lg}px) {
            .root {
              padding: 0 ${theme.spacing(5)}px;
            }
          }

          @media (max-width: ${theme.breakpoints.md}px) {
            .root {
              padding: 0 ${theme.spacing(3)}px;
            }
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses(Container);
