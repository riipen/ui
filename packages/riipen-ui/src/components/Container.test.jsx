import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import defaultTheme from "../themes/default";

import Container from "./Container";

describe("<Container>", () => {
  it("renders without errors", () => {
    let error;

    try {
      mount(<Container />);
    } catch (theError) {
      error = theError;
    }

    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<Container />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new Container().type.defaultProps;

      const wrapper = mount(<Container />);

      expect(wrapper.find("Container").props().maxWidth).toEqual(
        defaultProps.maxWidth
      );
    });
  });

  describe("children prop", () => {
    it("displays given children", () => {
      const child = <div>child</div>;

      const wrapper = mount(<Container>{child}</Container>);

      expect(wrapper.containsMatchingElement(child)).toBeTruthy();
    });
  });

  describe("classes prop", () => {
    it("sets custom class", () => {
      const classes = ["newClass"];

      const wrapper = mount(<Container classes={classes} />);

      expect(
        wrapper
          .find("Container")
          .childAt(0)
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to default classes prop with withClass decorator", () => {
      const wrapper = mount(<Container />);

      expect(wrapper.find("Container").props().classes).toEqual([
        "riipen",
        "riipen-container"
      ]);
    });
  });

  describe("maxWidth prop", () => {
    it("container renders with custom valid maxWidth", () => {
      const max = "md";

      const wrapper = mount(<Container maxWidth={max} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain(
        `${defaultTheme.breakpoints[max]}px`
      );
    });

    it("container renders at 100% width", () => {
      const max = false;

      const wrapper = mount(<Container maxWidth={max} />);

      expect(wrapper.find("JSXStyle").props().dynamic).toContain("100%");
    });

    it("throws an error with invalid maxWidth variant", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();
      const maxWidth = "s";

      mount(<Container maxWidth={maxWidth} />);

      expect(errors).toHaveBeenCalledTimes(1);
    });
  });
});
