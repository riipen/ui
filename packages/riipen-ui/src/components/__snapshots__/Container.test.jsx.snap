// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Container> renders correct snapshot 1`] = `
<withClasses(Container)>
  <Container
    classes={
      Array [
        "riipen",
        "riipen-container",
      ]
    }
    maxWidth="lg"
  >
    <div
      className="jsx-729050328 root riipen riipen-container"
      style={
        Object {
          "maxWidth": "1380px",
        }
      }
    >
      <JSXStyle
        dynamic={
          Array [
            30,
            1380,
            25,
            960,
            15,
          ]
        }
        id="3932038445"
      />
    </div>
  </Container>
</withClasses(Container)>
`;
